### PR TITLE
Add "Deal ticket size" to Large capital profile / Investor requirements 

### PR DESCRIPTION
**Pull request by paulgain**
**Add Deal ticket sizes to Large capital profile / Investor requirements**
Added 6 new checkboxes that allow a user to add the Deal ticket size.
 * Up to £49 million
 * £50-99 million
 * £100-249 million
 * £250-499 million
 * £500-999 million
 * £1 billion +

* Display what the user has selected within the profile read-only view.

**Refactoring of Large capital profile unit tests**
Some of the files where becoming unmanageable, therefore, I've broken
these out into smaller files.

https://uktrade.atlassian.net/browse/IPBETA-272

**Edit View**

edit-profile

**Read-only View**

read-only-profile

**Checklist**
- [ ] Has the branch been rebased to develop?
- [ ] Automated tests (Any of the following when applicable: Unit, Functional or Acceptance)
- [ ] Manual compatibility testing (Browsers: Chrome, Firefox, IE11, Safari)

### CODE REVIEW

File: src/apps/companies/apps/investments/large-capital-profile/styles.scss [hunk 1]
```diff
@@ -38,8 +38,23 @@ nav.govuk-tabs {
   }
 }
 
+.govuk-checkboxes__input +
+  .govuk-checkboxes__label::after {
+    width: 23px;
+    height: 12px;
+}
+
```
Comment (PanosParalakis) [inline src/apps/companies/apps/investments/large-capital-profile/styles.scss]: Do we really need to overwrite gov uk elements? If we really need to for our app maybe it is better to define those in a higher level so to be consistent across the app

File: src/apps/companies/apps/investments/macros/tasklist/tasklist.scss [hunk 2]
```diff
@@ -46,3 +46,7 @@
     }
   }
 }
+
```
Comment (PanosParalakis) [inline src/apps/companies/apps/investments/macros/tasklist/tasklist.scss]: You can use predefined gov uk spacing styles if you want to [https://design-system.service.gov.uk/styles/spacing/](https://design-system.service.gov.uk/styles/spacing/)

File: src/apps/companies/apps/investments/macros/details/details.scss [hunk 3]
```diff
@@ -29,7 +29,7 @@
   }
 
   .govuk-details__text {
-    margin-top: 10px;
+    margin-top: 20px;
```
Comment (PanosParalakis) [inline src/apps/companies/apps/investments/macros/details/details.scss]: You can use predefined gov uk spacing styles if you want to. See: [https://design-system.service.gov.uk/styles/spacing/](https://design-system.service.gov.uk/styles/spacing/)

File: src/apps/companies/apps/investments/utils/transformers.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    return obj
+  }
+}
+
+const ifStringAddToEmtpyArray = (obj) => {
+  return isString(obj) ? [obj] : obj
+}
+
+const sanitizeCheckboxes = (selection) => {
+  return selection === undefined ? [] : ifStringAddToEmtpyArray(selection)
```
Comment (yeahfro) [inline src/apps/companies/apps/investments/utils/transformers.js]: `castArray` of `lodash` can achieve the same thing

File: src/apps/companies/apps/investments/large-capital-profile/controllers/update.js [hunk 5]
```diff
@@ -10,6 +10,8 @@ const updateProfile = async (req, res, next) => {
   let body
   if (editing === INVESTOR_DETAILS) {
```
Comment (yeahfro) [inline src/apps/companies/apps/investments/large-capital-profile/controllers/update.js]: Another option for this is:

```
const transformer = {
  [INVESTOR_DETAILS]: transformInvestorDetails,
  [INVESTOR_REQUIREMENTS]: transformInvestorRequirements,
}
```

And then:

```
const body = transformer[editing](req.body)

```

Comment (paulgain) [reply src/apps/companies/apps/investments/large-capital-profile/controllers/update.js]: I like this. :thumbsup:

File: src/apps/companies/apps/investments/large-capital-profile/styles.scss (hunk 1 above)
Comment (paulgain) [reply src/apps/companies/apps/investments/large-capital-profile/styles.scss]: Thanks @PanosParalakis.

Looking deeper into this we don't need to set the width and  height. :wink: However, we need to override `box-sizing: border-box` and as you say we need to move it to a "higher level".

Data Hub uses a wildcard to apply `box-sizing: border-box` to every HTML element. 
https://github.com/uktrade/data-hub-frontend/blob/develop/assets/stylesheets/generic/_box-sizing.scss

This approach conflicts with GOV.UK checkboxes and as a consequence the tick within the checkbox appears clipped.

issue

 Overriding the generic approach by setting `box-sizing: content-box` back to its default rectifies the issue.

```
.govuk-checkboxes__input + .govuk-checkboxes__label::after {
  box-sizing: content-box;
}
```
fixed

The preferred approach would be to remove the wildcard, however, that will cause styling issues right across Data Hub. 

 @yeahfro

File: src/apps/companies/apps/investments/macros/details/details.scss (hunk 3 above)
Comment (paulgain) [reply src/apps/companies/apps/investments/macros/details/details.scss]: Personally I'm not a fan. You would need to know all the spacing units to understand what spacing was being applied.

File: src/apps/companies/apps/investments/utils/transformers.js (hunk 4 above)
Comment (paulgain) [reply src/apps/companies/apps/investments/utils/transformers.js]: I don't think I can use this, unless I'm mistaken, let me explain. 

**Background**
When a user checks a number of GOV.UK checkboxes and then POSTs those values to the server the request will be 1 of 3:

1.  User doesn't check any checkbox - server receives `undefined`
2. User checks a single checkbox - server receives a `String` (the id)
3. User selects more than one checkbox - server receives an `Array` of ids

We can only POST an array into the API which is either an empty array or a populated one.

Looking at the `castArray` docs... firstly, whatever is passed is added to an array including `undefined` which is not what we want. If it's undefined we need to return an empty `Array` not `[undefined]`.  Secondly, if it's a `String` then `castArray` is what we want, however, if it's an array of Strings we need to return that untouched.

Maybe I'm missing something? Could you provide a full example please. Thanks.

**Comment by jim68000** [comment]
This pull request **introduces 1 alert** when merging e41f979c31f72e9df46cccfaeeb9b95dc255634c into 6874fda8c57265bf020efe0d483da743bb959808 - [view on LGTM.com](https://lgtm.com/projects/g/uktrade/data-hub-frontend/rev/pr-a249d3ae9de92d5053ad129988ef57e4b46ff76d)

**new alerts:**

* 1 for Unvalidated dynamic method call

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by jim68000** [comment]
This pull request **introduces 1 alert** when merging f6847c5b0331ad7451ff6f5926ee2baeffca6a62 into 6874fda8c57265bf020efe0d483da743bb959808 - [view on LGTM.com](https://lgtm.com/projects/g/uktrade/data-hub-frontend/rev/pr-43095f45d9e69cdd37e54544f6899e12ea35d302)

**new alerts:**

* 1 for Unvalidated dynamic method call

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: src/apps/companies/apps/investments/utils/transformers.js (hunk 4 above)
Comment (yeahfro) [reply src/apps/companies/apps/investments/utils/transformers.js]: An option could also to be use `compact` in conjunction with `castArray`, but that's fair.

**Pull request merged**